contracts/token/erc721/abstract/ERC721Permit.sol

```diff
@@ -0,0 +1,216 @@
+                bytes32(BytesLib.slice(sig, 0, 32)),
+                bytes32(BytesLib.slice(sig, 32, 64))
+            );
+        } else if (sig.length == 65) {
+            // typical EDCSA Sig
+            recoveredSigner = ECDSA.recover(digest, sig);
+        } else {
+            revert InvalidSignature();
+        }
+
+        bool isNotZerothAddr = recoveredSigner != address(0);
+
+        if (!isNotZerothAddr) {
+            revert SignerCannotBeZerothAddress();
+        }
+
+        bool isApprovedOperator = _isApprovedOrOwner(recoveredSigner, tokenId);
+
+        bool isValidEOASig = isApprovedOperator && isNotZerothAddr;
+
+        if (!isValidEOASig) {
+            if (!_isValidERC1271Signature(getApproved(tokenId), digest, sig) &&
+                !_isValidERC1271Signature(ownerOf(tokenId), digest, sig)
+            ) {
+                revert InvalidSignature();
+            }
+        }
+
+        _approve(spender, tokenId);
+    }
+
+    /**
+     * @notice Returns the current nonce of a given token ID.
+     * @param tokenId The ID of the token for which to retrieve the nonce.
+     * @return Current nonce of the given token.
+     */
+    function nonces(
+        uint256 tokenId
+    ) external view returns (uint256) {
+        return _nonces[tokenId];
+    }
+
+    /**
+     * @notice Returns the domain separator used in the encoding of the signature for permits, as defined by EIP-712
+     * @return the bytes32 domain separator
+     */
+    function DOMAIN_SEPARATOR() external view override returns (bytes32) {
+        return _domainSeparatorV4();
+    }
+
+    /**
+     * @notice Builds the EIP-712 compliant digest for the permit.
+     * @param spender The address which is approved to spend the token.
+     * @param tokenId The ID of the token for which the permit is being generated.
+     * @param deadline The deadline until which the permit is valid.
+     * @return A bytes32 digest, EIP-712 compliant, that serves as a unique identifier for the permit.
+     */
+    function _buildPermitDigest(
+        address spender,
+        uint256 tokenId,
+        uint256 deadline
+    ) internal view returns (bytes32) {
+        return _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    _PERMIT_TYPEHASH,
+                    spender,
+                    tokenId,
+                    _nonces[tokenId],
+                    deadline
+                )
+            )
+        );
+    }
+
+    /**
+     * @notice Checks if a given signature is valid according to EIP-1271.
+     * @param spender The address which purports to have signed the message.
+     * @param digest The EIP-712 compliant digest that was signed.
+     * @param sig The actual signature bytes.
+     * @return True if the signature is valid according to EIP-1271, otherwise false.
+     */
+    function _isValidERC1271Signature(address spender, bytes32 digest, bytes memory sig) private view returns(bool) {
```

Comment:
Did you base this on an existing implementation or is there a doc that describes the logic here?

Reply:
The eip itself has an example implementation that calls the method directly on the wallet, however that does not let us revert with our own errors, the staticcall here is to give us more control over what behaviour we want if the signature verification fails within the eip1271 contract